src/iface/route.rs

```diff
@@ -0,0 +1,231 @@
+use managed::ManagedMap;
+use time::Instant;
+use core::ops::Bound;
+
+use {Error, Result};
+use wire::{IpCidr, IpAddress};
+#[cfg(feature = "proto-ipv4")]
+use wire::{Ipv4Address, Ipv4Cidr};
+#[cfg(feature = "proto-ipv6")]
+use wire::{Ipv6Address, Ipv6Cidr};
+
+/// A prefix of addresses that should be routed via a router
+#[derive(Debug, Clone, Copy)]
+pub struct Route {
+    pub via_router: IpAddress,
+    /// `None` means "forever".
+    pub preferred_until: Option<Instant>,
+    /// `None` means "forever".
+    pub expires_at: Option<Instant>,
+}
+
+impl Route {
+    /// Returns a route to 0.0.0.0/0 via the `gateway`, with no expiry.
+    #[cfg(feature = "proto-ipv4")]
+    pub fn new_ipv4_gateway(gateway: Ipv4Address) -> Route {
+        Route {
+            via_router: gateway.into(),
+            preferred_until: None,
+            expires_at: None,
+        }
+    }
+
+    /// Returns a route to ::/0 via the `gateway`, with no expiry.
+    #[cfg(feature = "proto-ipv6")]
+    pub fn new_ipv6_gateway(gateway: Ipv6Address) -> Route {
+        Route {
+            via_router: gateway.into(),
+            preferred_until: None,
+            expires_at: None,
+        }
+    }
+}
+
+/// A routing table.
+///
+/// # Examples
+///
+/// On systems with heap, this table can be created with:
+///
+/// ```rust
+/// use std::collections::BTreeMap;
+/// use smoltcp::iface::Routes;
+/// let mut routes = Routes::new(BTreeMap::new());
+/// ```
+///
+/// On systems without heap, use:
+///
+/// ```rust
+/// use smoltcp::iface::Routes;
+/// let mut routes_storage = [];
+/// let mut routes = Routes::new(&mut routes_storage[..]);
+/// ```
+#[derive(Debug)]
+pub struct Routes<'a> {
+    storage: ManagedMap<'a, IpCidr, Route>,
+}
+
+impl<'a> Routes<'a> {
+    /// Creates a routing tables. The backing storage is **not** cleared
+    /// upon creation.
+    pub fn new<T>(storage: T) -> Routes<'a>
+            where T: Into<ManagedMap<'a, IpCidr, Route>> {
+        let storage = storage.into();
+        Routes { storage }
+    }
+
+    /// Update the routes of this node.
+    pub fn update_routes<F: FnOnce(&mut ManagedMap<'a, IpCidr, Route>)>(&mut self, f: F) {
+        f(&mut self.storage);
+    }
+
+    /// Add a default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
+    ///
+    /// On success, returns the previous default gateway, if any.
+    #[cfg(feature = "proto-ipv4")]
+    pub fn add_default_ipv4_gateway(&mut self, gateway: Ipv4Address) -> Result<Option<Route>> {
```

Comment:
This adds a default IPv4 *route* (i.e. a single route is added), not *gateway* (which would imply that other routes might use the same gateway if none is specified). Also docstring and IPv6 needs to be changed.